[MRG+1] Add read_raw_gdf

cbrnr opened the pull request:
This adds `read_raw_gdf`, which is just a reference to `read_raw_edf`. I think that this makes it easier for people to load GDF files, since it is not immediately clear that `read_raw_edf` also loads GDF.

wmvanvliet commented:
In that case, should we also have a `read_raw_bdf` function? (BioSemi's native format, a quite common EEG machine)

cbrnr commented:
Good point, I think this would make sense. Every file type (file extension) should have its own reader, at least from a user-facing perspective.

agramfort commented:
no objection but I see this as something "nice to have". But if
somebody has the time sure.

cbrnr commented:
I'll do it as part of this PR.

agramfort commented on mne/io/edf/edf.py:
```diff
@@ -1078,87 +1165,169 @@ def _find_tal_idx(ch_names):
 
 
+def read_raw_gdf(input_fname, montage=None, eog=None, misc=None,
+                 stim_channel='auto', exclude=(), preload=False, verbose=None):
```
maybe not for this PR but I think we should move away such complex signature for raw_ readers for EEG.

I would advertise:

```
raw = read_raw_gdf(input_fname, stim_channel='auto', exclude=(), preload=False, verbose=None)
raw.set_montage(None)
raw.set_ch_types(dict(eog=None, misc=None))
```

any objection on the vision?

cbrnr replied:
I think this is a good idea. Let's do that in another PR.

cbrnr replied:
Then we can also remove the `stim_channel` argument, because it only sets the channel type and this can also be done afterwards. Or better yet, `mne.find_events` has a `stim_channel` argument that we could use to specify which channel is the stim channel (probably this already works).

larsoner replied:
Can you open an issue and tag 0.18 so we don't forget to remove before release? Otherwise we'll need a deprecation cycle if we forget to do it.

agramfort commented:
stim_channel is different as it should not be resampled the same way. For
EDF with different sfreq per channels it's necessary to keep it in the init.

>

cbrnr commented:
Ready - @wmvanvliet and/or @agramfort do you have time to review?

wmvanvliet commented on mne/io/edf/edf.py:
```diff
@@ -78,47 +79,47 @@ def find_edf_events(raw):
-                     and put in the raw.annotations attribute.
-                     Use :func:`mne.events_from_annotations` to obtain
-                     events for the annotations instead.
+        Defaults to 'auto', which means that the type of channels named
```
`type of channels named` --> `channels named`

cbrnr replied:
Done

wmvanvliet commented on mne/io/edf/edf.py:
```diff
@@ -78,47 +79,47 @@ def find_edf_events(raw):
-                     events for the annotations instead.
+        Defaults to 'auto', which means that the type of channels named
+        'status' or 'trigger' (case insensitive) is set to STIM. If str (or
+        list of str), the type of channels matching the name(s) is set to STIM.
```
`the type of channels matching the name(s) is` --> `the channels matching the name(s) are`

wmvanvliet commented on mne/io/edf/edf.py:
```diff
@@ -1078,87 +1165,250 @@ def _find_tal_idx(ch_names):
-                     and put in the raw.annotations attribute.
-                     Use :func:`mne.events_from_annotations` to obtain
-                     events for the annotations instead.
+        Defaults to 'auto', which means that the type of channels named
```
`type of channels` --> `channels`

wmvanvliet commented on mne/io/edf/edf.py:
```diff
@@ -1078,87 +1165,250 @@ def _find_tal_idx(ch_names):
-                     events for the annotations instead.
+        Defaults to 'auto', which means that the type of channels named
+        'status' or 'trigger' (case insensitive) is set to STIM. If str (or
+        list of str), the type of channels matching the name(s) is set to STIM.
```
`type of channels matching the name(s) is` --> `channels matching the name(s) are`

wmvanvliet commented on mne/io/edf/edf.py:
```diff
@@ -1078,87 +1165,250 @@ def _find_tal_idx(ch_names):
+        Defaults to 'auto', which means that the type of channels named
+        'status' or 'trigger' (case insensitive) is set to STIM. If str (or
+        list of str), the type of channels matching the name(s) is set to STIM.
+        If int (or list of ints), the type of channels corresponding to the
```
`type of channels` --> `channels`

wmvanvliet commented on mne/io/edf/edf.py:
```diff
@@ -78,47 +79,47 @@ def find_edf_events(raw):
+        Defaults to 'auto', which means that the type of channels named
+        'status' or 'trigger' (case insensitive) is set to STIM. If str (or
+        list of str), the type of channels matching the name(s) is set to STIM.
+        If int (or list of ints), the type of channels corresponding to the
```
`type of channels` --> `channels`

wmvanvliet commented on mne/io/edf/edf.py:
```diff
@@ -1078,87 +1165,250 @@ def _find_tal_idx(ch_names):
+        channels. Values should correspond to the electrodes in the file.
+        Default is None.
+    stim_channel : 'auto' | str | list of str | int | list of int
+        Defaults to 'auto', which means that the type of channels named
```
`type of channels` --> `channels`

etc.

wmvanvliet commented on mne/io/edf/edf.py:
```diff
@@ -1078,87 +1165,250 @@ def _find_tal_idx(ch_names):
+        channels. Values should correspond to the electrodes in the file.
+        Default is None.
+    stim_channel : 'auto' | str | list of str | int | list of int
+        Defaults to 'auto', which means that the type of channels named
```
`type of channels` --> `channels`

agramfort commented:
LGTM but I took a really fast look. I would suggest to wait also for @massich or @larsoner to have a look

cbrnr commented:
Thanks @wmvanvliet, I've changed the docstrings accordingly.

massich commented:
I'm AFK so I read in diagonal.

Since we are using pytest.mark.parametrize for fname we should also use it
for reader and avoid using the conditional inside the test because the call
is identical.

What I'm not sure is if it would be even better to have test_edf.py and
test_bdf.py.

Then if we do this, what should we do with the GDF?


On Fri, Feb 8, 2019, 11:36 Marijn van Vliet <notifications@github.com>
wrote:

> *@wmvanvliet* approved this pull request.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/5877#pullrequestreview-201492900>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGt-4zVocGwylCpAnaO-dyY0Y__BIdEFks5vLUUQgaJpZM4aTkl->
> .
>

larsoner commented on mne/io/edf/edf.py:
```diff
@@ -144,32 +144,32 @@ class RawEDF(BaseRaw):
     --------
-    mne.io.Raw : Documentation of attribute and methods.
+    mne.io.Raw : Documentation of attributes and methods.
+    mne.io.RawGDF : Raw object for GDF files.
```
Linking to `Raw` makes sense to get attributes, but we should link to `read_raw_*` functions for other readers.

cbrnr replied:
I'm not sure what you want me to do. Should I add `mne.io.read_raw_edf` here?

larsoner replied:
Yes and remove `RawGDF`

larsoner replied:
... and the description of `read_raw_edf` could mention "the preferred way to read EDF files", since we'd like people to use the functions instead of the classes where possible

cbrnr replied:
OK done.

larsoner commented on mne/io/edf/edf.py:
```diff
@@ -202,124 +198,97 @@ def __init__(self, input_fname, montage, eog=None, misc=None,
+    See Also
+    --------
+    mne.io.Raw : Documentation of attributes and methods.
+    mne.io.RawEDF : Raw object for EDF, EDF+, and BDF files.
```
Same

massich commented:
Ok. Tests are readable enough then, and if we change something in the code we can refractor the tests then.

agramfort commented:
@cbrnr you need to also edit this page : https://martinos.org/mne/dev/manual/io.html

cbrnr commented:
@larsoner what do we remove in 0.18? `stim_channel` will be removed in 0.19 AFAIK.

cbrnr commented:
@agramfort I've updated the IO manual (I've changed bits and pieces in other sections to improve readability, but I didn't go through the whole doc).

agramfort commented:
anyone else doing a pass and happy set it to MRG+2 so we can merge.

thx

cbrnr commented:
@larsoner can you please take a final look and merge if you're happy? I don't know what you meant by creating and tagging a new issue for 0.18 though, otherwise I've addressed all comments.

massich commented:
Thx @cbrnr and the others for the heavy lifting.
